macaw-aarch32/src/Data/Macaw/ARM/Semantics/TH.hs

```diff
@@ -414,79 +422,285 @@ unitValue =  M.Initial ARMDummyReg
 natReprFromIntTH :: Int -> Q Exp
 natReprFromIntTH i = [| knownNat :: M.NatRepr $(litT (numTyLit (fromIntegral i))) |]
 
+-- | A representation of an
+-- un-executed effectful generator action, where 'addr' is a generalized
+-- address (either a memory address, or a register number) and 'val' is the value
+-- to be written.
+data WriteAction f w where
+  -- | A single write action 
+  WriteAction :: forall f w
+               . (1 <= w)
+              => M.NatRepr w
+              -> f M.BoolType
+              -- ^ a guard which indicates whether or not this action should be committed
+              -> f (M.BVType (8 TL.* w))
```

Comment:
This is the value being written?  I assume that the location is now implicit because this will be run at a higher level where the destination of the write is apparent.

Reply:
oh, `w` is the number of bytes

Reply:
Yeah, `w` is the number of bytes to be written, and `addr` is the width of the "address" (i.e. 32 bits for memory accesses, 4 bits for GPRs and 8 bits for SIMDs). It doesn't appear here now because it instead appears as the key to the map, mapping addresses to write actions.